Source/Core/CameraLimiter.js

```diff
@@ -0,0 +1,301 @@
+        Check.defined('this.boundingObject.x', this.boundingObject.x);
+        Check.defined('this.boundingObject.y', this.boundingObject.y);
+        Check.defined('this.boundingObject.width', this.boundingObject.width);
+        Check.defined('this.boundingObject.height', this.boundingObject.height);
+        //>>includeEnd('debug');
+
+        var rect = this.boundingObject;
+
+        var rectMinimum = new Cartesian2(rect.x, rect.y);
+        var rectMaximum = new Cartesian2(rect.x + rect.width, rect.y + rect.height);
+        var returningPosition = new Cartesian2(position.x, position.y);
+
+        returningPosition.x = Math.max(rectMinimum.x, Math.min(rectMaximum.x, returningPosition.x));
+        returningPosition.y = Math.max(rectMinimum.y, Math.min(rectMaximum.y, returningPosition.y));
+
+        return new Cartesian3(returningPosition.x, returningPosition.y, position.z);
+    };
+
+    /**
+     * @private
+     */
+    CameraLimiter.prototype._closestLocationInBoundingSphere = function(position) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('this.boundingObject.radius', this.boundingObject.radius);
+        Check.defined('this.boundingObject.center', this.boundingObject.center);
+        //>>includeEnd('debug');
+
+        var center = this.boundingObject.center;
+        var radius = this.boundingObject.radius;
+
+        // to avoid dividing by zero check if same location as center.
+        if (position.equals(center)) {
+            return position;
+        }
+
+        // convert to cartographic and check if at appropriate height - sphere radius - if not then clarify
+        // var cartographicRep = Cartographic.fromCartesian(position);
+        // if (cartographicRep.height > )
+
+        var direction = Cartesian3.subtract(center, position, new Cartesian3());
+        Cartesian3.normalize(direction, direction);
+        Cartesian3.multiplyByScalar(direction, radius, direction);
+        var locationOnSphereInDirection = Cartesian3.add(direction, center, new Cartesian3());
+
+        var distanceToLocOnSphere = Cartesian3.distanceSquared(locationOnSphereInDirection, center);
+        var distanceToPosition = Cartesian3.distanceSquared(position, center);
+
+        return (distanceToPosition < distanceToLocOnSphere || distanceToPosition) ? position : locationOnSphereInDirection;
+    };
+
+    /**
+     * @private
+     */
+    CameraLimiter.prototype._closestLocationInOrientedBoundingBox = function(position) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('this.boundingObject.center', this.boundingObject.center);
+        Check.defined('this.boundingObject.halfAxes', this.boundingObject.halfAxes);
+        //>>includeEnd('debug');
+
+        // to avoid inverse of a zero matrix (since that doesn't exist).
+        if (this.boundingObject.halfAxes.equals(Matrix3.ZERO)) {
+            return (position.equals(Cartesian3.ZERO));
+        }
+
```

Comment:
@ggetz - could you review my math for here(_closestLocationInOrientedBoundingBox)

for closestLocationInOrientedBoundingBox what i'm currently doing is, since the halfAxes is scale of the item divided by two - to get the actual transformation, i'm scaling the overall by 2, using the inverse of this to transform back to object space (ie just a unit cube centered at the origin), calculating the appropriate min/maxes based on cube space, then converting back to worldspace for returned position.

for closestLocationInBoundingSphere -[my comments for me]- I'm "zeroing" the sphere by subtracting the center from the position so it's as if the position is relative to a sphere centered at zero - it seems i've currently forgotten to include a var storing this original centered location so that needs to be used instead of center for distToLocOnSphere, LocOnSphereInDir, and distToPosition.